### PR TITLE
Consistent etcd key path

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR lets the resources under the system-tenant space and non-system-tenant space have the same format of etcd keys. 

Previously, for the convenience of development, the resource of system tenant has a backward-compatible etcd key format: 
```
/registry/[resource_type]/[ns_name]/[resource_name]
```
while those of non-system tenant have a different format:
```
/registry/[resource_type]/[tenant-Name]/[ns_name]/[resource_name]
```

However, as the system-tenant etcd key prefix is the same as that of all-tenant etcd key prefix, it caused the problem that list the resources under system tenants actually will return the resources of every tenant.  This leads to problems like https://github.com/CentaurusInfra/arktos/issues/339, which hinders the development of other multi-tenancy features.

With this change, the etcd key of the resource of system tenant  is now in the following format: 
```
/registry/[resource_type]/system/[ns_name]/[resource_name]
```

#### 1. Check the etcd key path generated:
Check the etcd keys with the following command,
```
ETCDCTL_API=3 etcdctl get "" --prefix=true --keys-only
```

 and the tenant of "system" is shown in the etcd key paths as expected
image

#### 2. Verify that issue 339 is resolved
a. create a tenant 
```
kubectl create tenant x
```

b. Run Command to show the namespaces of system-tenant
```
 kubectl get namepsace --tenant system
```

As show below, only the namespaces under system-tenant will be shown:
image

Yet, before this change, the output will be as follows, where the namespaces under all the tenants will be returned:

image

c. The code change was verified by the multi-tenancy e2e test.

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -331,9 +331,7 @@ type Initializer struct {
 
 const (
 	// TenantAll is the default argument to specify on a context when you want to list or filter resources across all tenants
-	TenantAll string = ""
-	// Sometimes we need to explicitly indicate the operation targets at all the tenants, not the empty string of the above TenantAll
-	TenantAllExplicit string = "all"
+	TenantAll string = "all"
```
Comment (pdgetrf) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:334]: with this, "all" would be a reserved phrase and it should be checked against when regular user creates tenant

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 2]
```diff
@@ -253,15 +251,12 @@ func TenantKeyRootFunc(ctx context.Context, prefix string) string {
 // namespace-scope resource relative to the given prefix enforcing tenant/namespace rules. If the
 // context does not contain a tenant/namespace, it errors.
 func NamespaceKeyFunc(ctx context.Context, prefix string, name string) (string, error) {
-	var te, ns string
+	var tenant, ns string
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:254]: god yes.

File: staging/src/k8s.io/client-go/dynamic/fake/simple.go [hunk 3]
```diff
@@ -80,11 +80,15 @@ type dynamicResourceClient struct {
...
-	return &dynamicResourceClient{client: c, resource: resource}
+	return &dynamicResourceClient{client: c, resource: resource, namespace: metav1.NamespaceAll, tenant: metav1.TenantAll}
 }
 
 func (c *dynamicResourceClient) Namespace(ns string) dynamic.ResourceInterface {
-	return c.NamespaceWithMultiTenancy(ns, metav1.TenantSystem)
+	if ns != metav1.NamespaceAll {
+		return c.NamespaceWithMultiTenancy(ns, metav1.TenantSystem)
+	} else {
+		return c.NamespaceWithMultiTenancy(ns, metav1.TenantAll)
```
Comment (pdgetrf) [inline staging/src/k8s.io/client-go/dynamic/fake/simple.go:90]: nit: else { not really needed here.

**Review by pdgetrf — APPROVED** [review]
/lgtm

a few minor comments.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:334]: Correct. It is already checked here:

image

**Comment by chenqianfzh** [comment]
This PR resolves issues https://github.com/CentaurusInfra/arktos/issues/339 and https://github.com/CentaurusInfra/arktos/issues/435.

**Pull request merged**